master/internal/api_experiment.go

```diff
@@ -649,3 +652,59 @@ func (a *apiServer) MetricBatches(req *apiv1.MetricBatchesRequest,
 		}
 	}
 }
+
+func (a *apiServer) TrialsSnapshot(req *apiv1.TrialsSnapshotRequest,
+	resp apiv1.Determined_TrialsSnapshotServer) error {
+	experimentID := int(req.ExperimentId)
+	batchesProcessed := int(req.BatchesProcessed)
+	metricName := req.MetricName
+	metricType := req.MetricType
+	if metricType == apiv1.MetricType_METRIC_TYPE_UNSPECIFIED {
+		return errors.New("must specify a metric type")
+	}
+	if metricName == "" {
+		return errors.New("must provide metric name")
+	}
+
+	var startTime time.Time
+	for {
+		var response apiv1.TrialsSnapshotResponse
+
+		var newTrials []*apiv1.TrialsSnapshotResponse_Trial
+		var endTime time.Time
+		var err error
+		switch metricType {
+		case apiv1.MetricType_METRIC_TYPE_TRAINING:
+			newTrials, endTime, err = a.m.db.TrainingTrialsSnapshot(experimentID,
+				batchesProcessed, metricName, startTime)
+		case apiv1.MetricType_METRIC_TYPE_VALIDATION:
+			newTrials, endTime, err = a.m.db.ValidationTrialsSnapshot(experimentID,
+				batchesProcessed, metricName, startTime)
+		default:
+			panic("Invalid metric type")
+		}
+		if err != nil {
+			return errors.Wrapf(err,
+				"error fetching snapshots of metrics for %s metric %s in experiment %d at %d batches",
+				metricType, metricName, experimentID, batchesProcessed)
+		}
+		startTime = endTime
+
+		response.Trials = newTrials
+
+		if err := resp.Send(&response); err != nil {
+			return errors.Wrapf(err, "error sending batches recorded for metrics")
+		}
+
+		experiment, _ := a.getExperiment(experimentID)
```

Comment:
what do you think about a query for `getExperimentState`? just starting to call this a lot all over, and it has the full config which can be not tiny sometimes (100-150 lines of yaml, just looked one I had lying around to repro a bug was 1.8k).

Reply:
Yeah I was on the fence about that anyway, but if they can get that big that's really worth it. I've got a list going of things I want to improve about these endpoints now that I've done all 4 that were planned - I'll add this to the list of things to improve across all of them once the basics are in.